File: oauth.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"sync"
	"time"

	"golang.org/x/oauth2"
	"golang.org/x/oauth2/google"
)

// LoadClientSecretFromFile loads the client OAuth secret from the specified file
func LoadClientSecretFromFile(filePath string) (*oauth2.Config, error) {
	// Read the secret file
	b, err := os.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("error reading client secret file: %v", err)
	}

	// Configure the OAuth client
	config, err := google.ConfigFromJSON(b, "https://www.googleapis.com/auth/youtube.readonly")
	if err != nil {
		return nil, fmt.Errorf("error parsing client secret file: %v", err)
	}

	// Set the redirect URL
	config.RedirectURL = "http://localhost:8080/oauth2callback"

	return config, nil
}

// TokenFile returns the path to the token storage file
func TokenFile() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("failed to get user home directory: %w", err)
	}
	return filepath.Join(home, ".liveframe", "token.json"), nil
}

// TokenFromFile retrieves a token from a local file
func TokenFromFile(file string) (*oauth2.Token, error) {
	f, err := os.Open(file)
	if err != nil {
		return nil, fmt.Errorf("failed to open token file: %w", err)
	}
	defer f.Close()
	tok := &oauth2.Token{}
	err = json.NewDecoder(f).Decode(tok)
	if err != nil {
		return tok, fmt.Errorf("failed to decode token from file: %w", err)
	}
	return tok, nil
}

// SaveToken saves a token to a file
func SaveToken(path string, token *oauth2.Token) error {
	// Ensure directory exists
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0700); err != nil {
		return fmt.Errorf("failed to create token directory: %w", err)
	}

	f, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
	if err != nil {
		return fmt.Errorf("failed to open token file for writing: %w", err)
	}
	defer f.Close()

	if err := json.NewEncoder(f).Encode(token); err != nil {
		return fmt.Errorf("failed to encode token to file: %w", err)
	}
	return nil
}

// GetOAuthClient creates an HTTP server for OAuth flow and returns authorized client
// If forceAuth is true, it will start a new OAuth flow regardless of existing token
func GetOAuthClient(ctx context.Context, config *oauth2.Config, forceAuth bool) (*http.Client, error) {
	// Try to load token from file
	tokenPath, err := TokenFile()
	if err != nil {
		return nil, fmt.Errorf("failed to get token file path: %w", err)
	}

	// If not forcing auth, try to use existing token
	if !forceAuth {
		tok, err := TokenFromFile(tokenPath)
		if err == nil {
			// Token exists and was loaded successfully
			return config.Client(ctx, tok), nil
		}
	}

	// If forceAuth is true or token doesn't exist/is invalid - start OAuth flow
	log.Println("Starting new OAuth authentication flow")
	var tok *oauth2.Token
	codeChan := make(chan string)
	var wg sync.WaitGroup
	wg.Add(1)

	// Create a server mux for the HTTP server
	mux := http.NewServeMux()

	// Create an HTTP server to handle the OAuth callback
	server := &http.Server{
		Addr:    ":8080",
		Handler: mux,
	}

	// Set up the handler for the OAuth callback
	mux.HandleFunc("/oauth2callback", func(w http.ResponseWriter, r *http.Request) {
		code := r.URL.Query().Get("code")
		codeChan <- code

		// Display success message
		w.Header().Set("Content-Type", "text/html")
		fmt.Fprintf(w, "<h1>Authorization Successful</h1><p>You can close this window now.</p>")

		// Shutdown the server after a short delay
		go func() {
			time.Sleep(2 * time.Second)
			server.Shutdown(ctx)
			wg.Done()
		}()
	})

	// Start the HTTP server
	go func() {
		if err := server.ListenAndServe(); err != http.ErrServerClosed {
			log.Printf("HTTP server error: %v", err)
		}
	}()

	// Generate the authorization URL with maximum token lifetime
	// AccessTypeOffline provides a refresh token
	// ApprovalForce ensures we get a fresh refresh token by forcing the consent screen
	authURL := config.AuthCodeURL(
		"state-token",
		oauth2.AccessTypeOffline,
		oauth2.ApprovalForce,
	)

	// Open the URL in browser
	log.Printf("Opening browser for OAuth authorization: %s", authURL)
	if err := OpenURL(authURL); err != nil {
		log.Printf("Failed to open browser automatically: %v", err)
		fmt.Printf("Please open the following URL in your browser to authorize the application:\n%v\n", authURL)
	} else {
		log.Println("Browser opened for authorization")
	}

	// Wait for the authorization code
	code := <-codeChan

	// Exchange the code for a token
	tok, err = config.Exchange(ctx, code)
	if err != nil {
		return nil, fmt.Errorf("error exchanging code for token: %v", err)
	}

	// Save the token for future use
	if err := SaveToken(tokenPath, tok); err != nil {
		log.Printf("Warning: Failed to save token: %v", err)
	}

	// Wait for the server to shutdown
	wg.Wait()

	return config.Client(ctx, tok), nil
}

File: src/oauth.rs
use std::collections::HashMap;
use std::fs::{self, File};
use std::io::{Read, Write};
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use std::time::Duration;

use chrono::Utc;
use dirs::home_dir;
use hyper::service::{make_service_fn, service_fn};
use hyper::{Body, Request, Response, Server, StatusCode};
use log::{debug, info, warn};
use oauth2::{
    AuthUrl, AuthorizationCode, ClientId, ClientSecret, CsrfToken, PkceCodeChallenge,
    RedirectUrl, RefreshToken, Scope, TokenResponse, TokenUrl,
};
use oauth2::basic::BasicClient;

use crate::models::{ClientSecrets, OAuthState, TokenInfo};

// Function to get OAuth token (either from file or through auth flow)
pub async fn get_oauth_token() -> std::result::Result<TokenInfo, Box<dyn std::error::Error>> {
    // Check if token file exists
    let token_path = get_token_path()?;
    if token_path.exists() {
        info!("Found existing token file, loading...");
        let mut file = File::open(&token_path)?;
        let mut contents = String::new();
        file.read_to_string(&mut contents)?;
        
        let token_info: TokenInfo = serde_json::from_str(&contents)?;
        
        // If token is not expired, return it
        if Utc::now() < token_info.expiry {
            debug!("Token is still valid, using existing token");
            return Ok(token_info);
        }
        
        // If token is expired, try to refresh it
        info!("Token expired, refreshing...");
        match refresh_token(&token_info.refresh_token).await {
            Ok(new_token) => return Ok(new_token),
            Err(e) => warn!("Failed to refresh token: {}, starting new auth flow", e),
        }
    }
    
    // If no valid token exists, start OAuth flow
    info!("Starting OAuth authentication flow...");
    let token_info = oauth_flow().await?;
    
    // Save token to file
    save_token(&token_info)?;
    debug!("Token saved to file");
    
    Ok(token_info)
}

// Function to get the path to the token file
pub fn get_token_path() -> std::result::Result<PathBuf, Box<dyn std::error::Error>> {
    let mut path = home_dir().ok_or("Could not find home directory")?;
    path.push(".liveframe");
    
    // Create directory if it doesn't exist
    if !path.exists() {
        fs::create_dir_all(&path)?;
    }
    
    path.push("token.json");
    Ok(path)
}

// Function to get the path to the client secrets file
pub fn get_secrets_path() -> std::result::Result<PathBuf, Box<dyn std::error::Error>> {
    let mut path = home_dir().ok_or("Could not find home directory")?;
    path.push(".liveframe");
    path.push("secret.json");
    
    if !path.exists() {
        return Err("Client secrets file not found at ~/.liveframe/secret.json".into());
    }
    
    Ok(path)
}

// Function to save token to file
pub fn save_token(token_info: &TokenInfo) -> std::result::Result<(), Box<dyn std::error::Error>> {
    let token_path = get_token_path()?;
    let json = serde_json::to_string_pretty(token_info)?;
    let mut file = File::create(token_path)?;
    file.write_all(json.as_bytes())?;
    Ok(())
}

// Function to load client secrets
pub fn load_client_secrets() -> std::result::Result<ClientSecrets, Box<dyn std::error::Error>> {
    let secrets_path = get_secrets_path()?;
    let mut file = File::open(secrets_path)?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;
    
    let secrets: ClientSecrets = serde_json::from_str(&contents)?;
    Ok(secrets)
}

// Function to perform OAuth flow
pub async fn oauth_flow() -> std::result::Result<TokenInfo, Box<dyn std::error::Error>> {
    // Load client secrets
    info!("Loading client secrets...");
    let secrets = load_client_secrets()?;
    
    // Create OAuth client
    debug!("Creating OAuth client...");
    let client = BasicClient::new(
        ClientId::new(secrets.installed.client_id),
        Some(ClientSecret::new(secrets.installed.client_secret)),
        AuthUrl::new(secrets.installed.auth_uri)?,
        Some(TokenUrl::new(secrets.installed.token_uri)?),
    )
    .set_redirect_uri(RedirectUrl::new("http://localhost:8080".to_string())?);
    
    // Generate PKCE challenge
    debug!("Generating PKCE challenge...");
    let (pkce_challenge, pkce_verifier) = PkceCodeChallenge::new_random_sha256();
    
    // Generate the authorization URL
    let (auth_url, csrf_state) = client
        .authorize_url(CsrfToken::new_random)
        .add_scope(Scope::new("https://www.googleapis.com/auth/youtube.readonly".to_string()))
        .set_pkce_challenge(pkce_challenge)
        .url();
    
    info!("Open this URL in your browser to authorize the application:");
    info!("{}", auth_url);
    
    // Create a shared state for the callback server
    let state = Arc::new(Mutex::new(OAuthState {
        auth_code: None,
        csrf_state: csrf_state.secret().clone(),
        pkce_verifier: Some(pkce_verifier),
    }));
    
    // Start the HTTP server for the OAuth callback
    info!("Starting OAuth callback server on http://localhost:8080");
    let state_clone = state.clone();
    let make_service = make_service_fn(move |_| {
        let state = state_clone.clone();
        async move {
            Ok::<_, hyper::Error>(service_fn(move |req| {
                let state = state.clone();
                async move { handle_oauth_callback(req, state).await }
            }))
        }
    });
    
    let addr = ([127, 0, 0, 1], 8080).into();
    let server = Server::bind(&addr).serve(make_service);
    
    // Clone the state for use in the async block
    let state_for_server = state.clone();
    
    // Run the server with a timeout
    debug!("Waiting for authorization callback (timeout: 2 minutes)...");
    let server_with_timeout = async move {
        let server_future = server.with_graceful_shutdown(async {
            // Wait for the auth code to be received
            for i in 0..120 {  // Wait up to 2 minutes
                {
                    let state_guard = state_for_server.lock().unwrap();
                    if state_guard.auth_code.is_some() {
                        debug!("Authorization code received");
                        break;
                    }
                }
                if i % 30 == 0 && i > 0 {
                    debug!("Still waiting for authorization... ({} seconds elapsed)", i);
                }
                tokio::time::sleep(Duration::from_secs(1)).await;
            }
        });
        
        server_future.await
    };
    
    // Run the server and wait for it to complete
    server_with_timeout.await?;
    
    // Get the authorization code from the state
    let auth_code = {
        let state_guard = state.lock().unwrap();
        state_guard.auth_code.clone().ok_or("No authorization code received")?
    };
    
    // Take the PKCE verifier from the state
    let pkce_verifier = {
        let mut state_guard = state.lock().unwrap();
        state_guard.pkce_verifier.take().ok_or("PKCE verifier not found")?
    };
    
    // Exchange the authorization code for an access token
    info!("Exchanging authorization code for access token...");
    let token_result = client
        .exchange_code(AuthorizationCode::new(auth_code))
        .set_pkce_verifier(pkce_verifier)
        .request_async(oauth2::reqwest::async_http_client)
        .await?;
    
    // Create token info
    debug!("Creating token info with expiry time");
    let token_info = TokenInfo {
        access_token: token_result.access_token().secret().clone(),
        refresh_token: token_result.refresh_token()
            .ok_or("No refresh token received")?
            .secret()
            .clone(),
        expiry: Utc::now() + chrono::Duration::seconds(token_result.expires_in().unwrap_or_default().as_secs() as i64),
    };
    
    info!("OAuth flow completed successfully");
    Ok(token_info)
}

// Function to handle OAuth callback
pub async fn handle_oauth_callback(
    req: Request<Body>,
    state: Arc<Mutex<OAuthState>>,
) -> std::result::Result<Response<Body>, hyper::Error> {
    let uri = req.uri();
    let query = uri.query().unwrap_or("");
    
    let params: HashMap<_, _> = url::form_urlencoded::parse(query.as_bytes())
        .into_owned()
        .collect();
    
    let mut response = Response::new(Body::empty());
    
    if let (Some(code), Some(received_state)) = (params.get("code"), params.get("state")) {
        // Verify CSRF state
        let expected_state = {
            let state_guard = state.lock().unwrap();
            state_guard.csrf_state.clone()
        };
        
        if received_state == &expected_state {
            // Store the authorization code
            {
                let mut state_guard = state.lock().unwrap();
                state_guard.auth_code = Some(code.clone());
            }
            
            *response.body_mut() = Body::from(
                "Authorization successful! You can close this window and return to the application.",
            );
        } else {
            *response.status_mut() = StatusCode::BAD_REQUEST;
            *response.body_mut() = Body::from("Invalid state parameter");
        }
    } else {
        *response.status_mut() = StatusCode::BAD_REQUEST;
        *response.body_mut() = Body::from("Missing code or state parameter");
    }
    
    Ok(response)
}

// Function to refresh OAuth token
pub async fn refresh_token(refresh_token: &str) -> std::result::Result<TokenInfo, Box<dyn std::error::Error>> {
    // Load client secrets
    debug!("Loading client secrets for token refresh...");
    let secrets = load_client_secrets()?;
    
    // Create OAuth client
    let client = BasicClient::new(
        ClientId::new(secrets.installed.client_id),
        Some(ClientSecret::new(secrets.installed.client_secret)),
        AuthUrl::new(secrets.installed.auth_uri)?,
        Some(TokenUrl::new(secrets.installed.token_uri)?),
    );
    
    // Exchange the refresh token for a new access token
    info!("Exchanging refresh token for new access token...");
    let token_result = client
        .exchange_refresh_token(&RefreshToken::new(refresh_token.to_string()))
        .request_async(oauth2::reqwest::async_http_client)
        .await?;
    
    // Create token info
    let token_info = TokenInfo {
        access_token: token_result.access_token().secret().clone(),
        refresh_token: token_result.refresh_token()
            .map(|rt| rt.secret().clone())
            .unwrap_or_else(|| refresh_token.to_string()),
        expiry: Utc::now() + chrono::Duration::seconds(token_result.expires_in().unwrap_or_default().as_secs() as i64),
    };
    
    // Save the new token
    debug!("Saving refreshed token to file...");
    save_token(&token_info)?;
    info!("Token refreshed successfully");
    
    Ok(token_info)
}